### DISCUSSION TITLE
How would I send analytics data forwarded to my backend to Mixpanel, as opposed to sending data directly from the browser?

### DISCUSSION THREAD

**Discussion by hirehamir** [mixpanel/docs-legacy#589] [General]
I'm curious whether I'd be able to obfuscate the URL I'm sending analytics data to, regardless of whether or not I'm using a proxy, by sending Mixpanel analytics to a my backend, which would then forward the analytics data to a proxy, or directly to Mixpanel.

This way end-users on desktop don't see the analytics URL in the Network tab of their browser DevTools.

**Comment by austinpray-mixpanel** [comment]
I think https://github.com/mixpanel/tracking-proxy meets your requirements. Is this what you are looking for? cc @ak--47

**Reply by hirehamir** [reply]
My understanding is that end-users can still see requests to a proxy in their browser's DevTools section, since the requests still come from the browser.

**Reply by hirehamir** [reply]
I'm wondering if I can forward requests to `/api/something` in my frontend, using a framework like Next.js, and then my server endpoint at `/api/something` can actually forward any necessary information to Mixpanel to track analytics.

**Reply by enague** [reply]
I know AK from the Mixpanel team just put out this video mainly for ad blockers, but it goes through the same process of using a proxy: https://www.youtube.com/watch?v=8Pv6tmRfqr8. Could you check out the video and see if it helps? 

Your second option may work, but I have pretty limited knowledge for Next.js, so I would not be so sure. Could try it out though!

**Reply by enague** [reply]
Feel free to ask your question in our Slack community as well here: https://www.mixpanel.com/community-slack

**Reply by ak--47** [reply]
👋 @hamirmahal  

if you don't want end users to be able to see any analytics requests in the browser devtools, the only reliable solution is [server-side tracking](https://docs.mixpanel.com/docs/tracking/how-tos/effective-server-side-tracking) ... we have a [node.js SDK](https://docs.mixpanel.com/docs/tracking/reference/nodejs) that can be used with Next.js

if you  created your own server-side endpoint to call [the `/track` mixpanel API](https://developer.mixpanel.com/reference/track-event) and passed valid `{event: "", properties:{}}` records... your end-users will see the request in browser dev tools.

if the request originates from the client, the client can see the request... 

what's the use case here?

**Reply by hirehamir** [reply]
Hello, @ak--47!

**Reply by hirehamir** [reply]
I may take a look at https://docs.mixpanel.com/docs/tracking/how-tos/effective-server-side-tracking and https://docs.mixpanel.com/docs/tracking/reference/nodejs.

**Reply by hirehamir** [reply]
The use case is just that I was concerned any user, upon viewing the Mixpanel API endpoint in the browser dev tools, can send arbitrary events and incorrect analytics data to that endpoint, since they can see what it is.

**Reply by ak--47** [reply]
@hamirmahal  

i see... you're worried about people spamming your analytics endpoint? that's fair. 

if you move all the API calls to the server, then the client will not see anything... that's one advantage to server-side tracking...  however, a "server-side only: strategy may  lose the ability to "track" actions on the client which are not necessarily triggering requests to the server (like scrolls, or certain page clicks) so i understand why this might not work for your use case.

people have expressed this concern before to me (many times) ... and it's a problem not just for mixpanel, but for every analytics product or service which exposes a "client side" endpoint for logging... if it's visible to the client, it could be spammed or abused... strategies like CSRF and rate-limiting tokens can be introduced, but that end up being nontrivial and more investment than it's worth.

i can say this: i've worked with thousands of customers (8+ yrs @ Mixpanel) and I have only seen the "client-side event spamming" behavior two or three times. it really doesn't happen frequently.

in the (rare) case this does happen, mixpanel offers governance tooling (lexicon) to drop certain events at ingestions, mark things as unexpected, etc... 

https://docs.mixpanel.com/docs/data-governance/lexicon#dropping-and-hiding-data

**Reply by hirehamir** [reply]
@ak--47

Oh, dang. That's really interesting. Thanks for the detailed response!

**Reply by ak--47** [reply]
@hamirmahal we now support configurable endpoints... so you can alias `/track` to `/whatever`:

https://github.com/mixpanel/mixpanel-js/releases/tag/v2.48.0
